### ISSUE TITLE
dropout method

### ISSUE THREAD

**Issue by miziworld** [espnet/espnet#4123]
Hello, Always appreciate quick replies and project development.
I'm using the ASR system in espnet1, 
and I know that I can use dropout for training with the pruning technique provided by pytorch.
We know that to perform efficient pruning, we need to repeat the pruning->training->pruning->training process.

1.To do this, when running run.sh, after training a few epochs with the dropout layer added, change the option from checkpoint to resume option to dropout=0.0, is it possible to train by changing the dropout ratio again after training?
(ex, dropout=0.1, ->dropout=0.0->dropout=0.2->dropout=0.0...)

2. Also, the dropout function provided by PyTorch is a function that randomly cuts out neurons. Is there a way to remove the weight when the weight value is 0 or a value close to 0?

**Comment by sw005320** [comment]
> 1.To do this, when running run.sh, after training a few epochs with the dropout layer added, change the option from checkpoint to resume option to dropout=0.0, is it possible to train by changing the dropout ratio again after training? (ex, dropout=0.1, ->dropout=0.0->dropout=0.2->dropout=0.0...)

I have not tested this, but I believe it should work by changing the dropout config when you resume training.

> 2. Also, the dropout function provided by PyTorch is a function that randomly cuts out neurons. Is there a way to remove the weight when the weight value is 0 or a value close to 0?

We don't support such a function.
You may need to implement it by yourself.

**Comment by miziworld** [comment]
Ok, I'll try it.
But, when dropout(=0.2) was applied, training and decoding were performed, the number of parameters or model size did not decrease compared to when no dropout was applied. Why is that?
![image](https://user-images.githubusercontent.com/17849068/156725565-35336b3a-4228-4b7d-bf1c-4caf0a527b2e.png)

**Comment by csukuangfj** [comment]
> But, when dropout(=0.2) was applied, training and decoding were performed, the number of parameters or model size did not decrease compared to when no dropout was applied. Why is that?

Why do you expect a decrease?

**Comment by miziworld** [comment]
As you know, we know that by applying dropout, we can reduce the number of wieghts and the size of the model by eliminating the layered intermediate weights.
Is dropout applied in espnet a different feature?

**Comment by csukuangfj** [comment]
From the paper [Dropout: A Simple Way to Prevent Neural Networks from Overfitting](https://jmlr.org/papers/v15/srivastava14a.html)

> Dropout is a technique for addressing this problem. The key idea is to randomly drop units (along with their connections) from the neural network during training

It does not mention that dropout is able to reduce the number of weights of a model.

**Comment by miziworld** [comment]
Oh, I think I've misunderstood the concept of pruning. Thank you for your kind reply.
So, is there any way I can use pruning to speed up decoding in espnet?
Also, can you suggest me what I can do besides adjusting the beam-size to speed up the decoding in espnet1?
